sql-plugin/src/main/scala/com/nvidia/spark/rapids/PrioritySemaphore.scala

```diff
@@ -56,17 +54,28 @@ class PrioritySemaphore[T](val maxPermits: Int)(implicit ordering: Ordering[T])
 
   def acquire(numPermits: Int, priority: T): Unit = {
     lock.lock()
+    val condition = lock.newCondition()
+    val info = ThreadInfo(priority, condition, numPermits)
     try {
-      val condition = lock.newCondition()
-      while (!canAcquire(numPermits)) {
-        waitingQueue.enqueue(ThreadInfo(priority, condition))
-        condition.await()
+      if (canAcquire(numPermits)) {
```

Comment:
This needs the priority inversion check that matches the logic in tryAcquire.  Ideally this should be implemented in terms of tryAcquire, could refactor to have something like tryAcquireLockHeld that both tryAcquire and acquire use, or change `canAcquire` to take a priority parameter and move that logic into there.

Reply:
Oh good point

Reply:
I think I would just have `acquire` begin by calling `tryAcquire`, if true, done, else, the remaining queue logic.

I think leaving `canAcquire` as is makes sense because it's used by `release` in a scenario where we already know as we iterate the queue that the priority is correct

Reply:
```
    if (!tryAcquire(numPermits, priority)) {
      lock.lock()
      val condition = lock.newCondition()
      val info = ThreadInfo(priority, condition, numPermits)
      try {
        waitingQueue.add(info)
        while (!info.signaled) {
          info.condition.await()
        }
      } catch {
        case e: Exception =>
          waitingQueue.remove(info)
          if (info.signaled) {
            release(numPermits)
          }
          throw e
      } finally {
        lock.unlock()
      }
    }
    ```

Reply:
When you say `LockHeld` I guess you are referring to the `ReentrantLock` not the sempahore itself. I think it's fine to lock it twice as would happen here, but if it's preferred for whatever reason I can factor it out as you said.